src/d3a/models/state.py

```diff
@@ -72,14 +73,50 @@ class ProsumptionInterface(StateInterface, ABC):
     def __init__(self):
         # Actual energy consumed/produced by the device at specific market slots
         self._energy_measurement_kWh: Dict[DateTime, float] = {}
+        self._unsettled_deviation_kWh: Dict[DateTime, float] = {}
+        self._forecast_measurement_deviation_kWh: Dict[DateTime, float] = {}
+
+    def _calculate_unsettled_energy_kWh(self, measured_energy_kWh, time_slot):
+        """Return the unsettled energy (produced or consumed) in kWh."""
+        return 0.0
 
     def set_energy_measurement_kWh(self, energy_kWh: float, time_slot: DateTime) -> None:
         """Set the actual energy consumed/produced by the device in the given market slot."""
         self._energy_measurement_kWh[time_slot] = energy_kWh
+        self._forecast_measurement_deviation_kWh[time_slot] = self._calculate_unsettled_energy_kWh(
+            energy_kWh, time_slot)
+        self._unsettled_deviation_kWh[time_slot] = \
+            abs(self._forecast_measurement_deviation_kWh[time_slot])
 
     def get_energy_measurement_kWh(self, time_slot: DateTime) -> float:
         """Get the actual energy consumed/produced by the device in the given market slot."""
-        return self._energy_measurement_kWh.get(time_slot, None)
+        return self._energy_measurement_kWh.get(time_slot)
+
+    def get_forecast_measurement_deviation_kWh(self, time_slot: DateTime) -> float:
+        """Get the energy deviation of forecasted energy from measurement by the device in
+        the given market slot. Negative value means that the deviation is beneficial to the
+        grid (and can be posted as an offer), positive value means that the deviation is
+        detrimental to the grid (and can be posted as a bid)"""
+        return self._forecast_measurement_deviation_kWh.get(time_slot)
+
+    def should_post_bid(self, time_slot: DateTime) -> bool:
```

Comment:
The `should_post_bid` and `should_post_offer` methods should not belong to the `ProsumptionInterface`. As stated by the docstring, `ProsumptionInterface` is an
> Interface with common methods/variables shared by consumption and production devices.

The `should_post_offer` can't be used by consumption devices, and `should_post_bid` can't be used by production devices, so they would rather belong to `ConsumptionState` and `ProductionState`.

Reply:
I thought I had answered here, sorry. What you say makes sense only for the spot market actually, and not for the settlement market. All assets can post both bids and offers on the settlement market, according to the deviation from the forecasted energy (for instance, if a load consumed less than it bought, then it has a deviation on the negative side, which then can sell in the settlement market. Same for a PV that overproduces)
One improvement could be to rename these to can_post_settlement_bid and can_post_settlement_offer, which I will add in the context of 3568 in order to not block Hannes' PR

Reply:
Thanks @spyrostz! I have the feeling that this complicates a bit the understanding of the states on an intuitive level: it feels a bit counter-intuitive to have bid-related methods on PVs (and offer-related methods on Loads). The re-naming that you suggest would definitely help to make the difference more visible. Thanks again.